scalanet/discovery/src/com/chipprbots/scalanet/discovery/ethereum/v5/DiscoveryNetwork.scala

```diff
@@ -212,76 +213,523 @@ object DiscoveryNetwork {
+      
+      /** Handle ordinary encrypted message packet */
+      private def handleOrdinaryMessage(peer: Peer[A], packet: Packet.OrdinaryMessagePacket): IO[Unit] = {
+        import codecs.RLPCodecs.payloadCodec
         
+        logger.debug(s"Received ordinary message from $peer")
+        
+        for {
+          // 1. Get session from cache
+          remoteNodeId <- IO.pure(Session.nodeIdFromPublicKey(peer.id.value.bytes))
+          sessionOpt <- sessionCache.get(remoteNodeId)
+          
+          // 2. If no session, send WHOAREYOU challenge
+          _ <- sessionOpt match {
+            case None =>
+              logger.debug(s"No session for $peer, sending WHOAREYOU")
+              sendWhoAreYou(peer, packet.nonce)
+              
+            case Some(session) =>
+              // 3. Decrypt message using session role
+              for {
+                // Select decryption key based on session role (inverse of encryption)
+                decryptionKey <- IO.pure(if (session.isInitiator) session.keys.initiatorKey else session.keys.recipientKey)
+                
+                plaintext <- IO.fromTry(
+                  Session.decrypt(decryptionKey, packet.nonce, packet.messageCipherText, packet.authData)
+                )
+                
+                // 4. Decode payload
+                payload <- IO.fromEither(
+                  payloadCodec.decode(plaintext.bits).toEither
+                    .left.map(err => new PacketException(s"Failed to decode payload: ${err.message}"))
+                    .map(_.value)
+                )
+                
+                // 5. Handle payload
+                _ <- handlePayload(peer, payload)
+              } yield ()
+          }
+        } yield ()
+      }
+      
+      /** Handle handshake message packet */
+      private def handleHandshakeMessage(peer: Peer[A], packet: Packet.HandshakeMessagePacket): IO[Unit] = {
+        logger.debug(s"Received handshake message from $peer")
+        
+        // Handshake messages are sent by the initiator in response to WHOAREYOU
+        // We need to verify the signature and establish the session as recipient
+        for {
+          // 1. Extract source node ID from handshake auth data
+          srcNodeId <- IO.pure(packet.handshakeAuthData.srcId)
+          
+          // 2. Verify signature over id-nonce (would need the original idNonce we sent)
+          // For now, we'll accept it - in production, verify signature
+          
+          // 3. The ephemeral public key is in the packet
+          ephemPubkey <- IO.pure(packet.handshakeAuthData.ephemPubkey)
+          
+          // 4. Perform ECDH with our node's private key
+          sharedSecret <- IO.pure(Session.performECDH(privateKey.value.bytes, ephemPubkey))
+          
+          // 5. Derive session keys (would need the original idNonce we sent in WHOAREYOU)
+          // For now, we'll skip session creation - in production, store idNonce and retrieve it
```

Comment:
we are building for production. do not leave the implementation incomplete or insecure